### PR TITLE
Introduce ImmutableSnapshot strategy to create snapshots using timestamps

### PR DESCRIPTION
**Pull request by rpalcolea**
This introduces `immutableSnapshot` task/strategy to generate snapshot versions that are meant to be immutable and should be unique via UTC timestamp. Ex. `1.2.0-snapshot.20190705210556893+abcdef`  

Pattern is  `<major>.<minor>.<patch>-snapshot.<timestamp>+<hash>` and timestamp format is `YYYYMMddHHmmssSSS`

This change doesn't replace `devSnapshot` and default strategy to prevent breaking changes for consumers.

We should be able to override the default strategy from `devSnapshot` to `immutableSnapshot` via `ReleasePluginExtension`:

```
release {
                defaultVersionStrategy = nebula.plugin.release.NetflixOssStrategies.IMMUTABLE_SNAPSHOT(project)
            }
```

### CODE REVIEW

**Comment by rspieldenner** [comment]
if I have `1.2.3-snapshot.2019070511000000+abcdef` and `1.2.3-rc.1` which will I get if I ask for `1.+`?

**Comment by rpalcolea** [comment]
> if I have `1.2.3-snapshot.2019070511000000+abcdef` and `1.2.3-rc.1` which will I get if I ask for `1.+`?

@rspieldenner , this is actually a good question. This new immutable snapshot suggests to use the snapshot explicitly instead of relying on things like `1.+`. Because in semver identifiers consisting of only digits are compared numerically and identifiers with letters or hyphens are compared lexically in ASCII sort order,  `1.2.3-rc.1` would always be first.

This is different from the usual behavior with `-dev.`

I think this should be considered as a breaking change in order to unify versioning with other technologies that are Semver based, like NPM.

I believe snapshots should be always consumed explicitly and not via dynamic versions (`1.+`) but that's just my opinion.

Also, I think is worth mentioning that generating a snapshot from a branch where latest tag is a RC one, ends up in something like this: ` 1.2.3-rc.1.dev.0+cd60fe6` which would resolve to `1.2.3-rc.1` if you do `1.+` since `1.2.3-rc.1.dev.0+cd60fe6 < 1.2.3-rc.1`

**Comment by rspieldenner** [comment]
`1.2.3-rc.1.dev.0+cd60fe6` was an accident of how I implemented this if we can fix to always have snapshots be `-dev.#` or `-snapshot.timestamp` that would be better

File: README.md [hunk 1]
```diff
@@ -73,9 +73,10 @@ All tasks will trigger gradle-git's release task which is configured to depend o
...
+* immutableSnapshot - Sets the version to the appropriate `<major>.<minor>.<patch>-snapshot.<timestamp>+<hash>`, does not create a tag. Where `timestamp` is UTC time in `YYYYMMddHHmmssSSS` format, ex. `20190705210556893`  and `hash` is the git hash of the current commit.  If releasing a immutableSnapshot from a branch not listed in the `releaseBranchPatterns` and not excluded by `excludeBranchPatterns` the version will be `<major>.<minor>.<patch>-snapshot.<timestamp>+<branchname>.<hash>`
 * snapshot - Sets the version to the appropriate `<major>.<minor>.<patch>-SNAPSHOT`, does not create a tag.
 
-Use of devSnapshot vs snapshot is a project by project choice of whether you want maven style versioning (snapshot) or unique semantic versioned snapshots (devSnapshot).
+Use of immutableSnapshot vs devSnapshot vs snapshot is a project by project choice of whether you want maven style versioning (snapshot) or unique semantic versioned snapshots (devSnapshot) or unique semantic versioned snapshots that guaranteed uniqueness (immutableSnapshot).
```
Comment (OdysseusLives) [inline README.md]: Can you adjust the wording: `or unique semantic versioned snapshots (devSnapshot)` to `or semantic versioned snapshots (devSnapshot)`

File: README.md [hunk 2]
```diff
@@ -165,20 +173,22 @@ The plugin assumes Git root is in the same location as Gradle root. If this isn'
...
+Built with OpenJDK8
+Tested with OpenJDK8
 
 | Gradle Version | Works |
 | :------------: | :---: |
-| 2.13           | yes   |
-| 3.3            | yes   |
-| 3.4.1          | yes   |
-| 3.5            | yes   |
+| 5.0            | yes   |
```
Comment (OdysseusLives) [inline README.md]: YAY

File: src/integTest/groovy/nebula/plugin/release/ReleasePluginIntegrationSpec.groovy [hunk 3]
```diff
@@ -406,6 +485,20 @@ class ReleasePluginIntegrationSpec extends GitVersioningIntegrationSpec {
...
+        git.push(all: true)
+        git.branch.change(name: 'feature/testexample', startPoint: 'origin/feature/testexample')
+        git.checkout(branch: 'feature/testexample')
+
+        when:
+        def version = inferredVersionForTask('immutableSnapshot')
+
+        then:
+        version.toString().startsWith("0.1.0-snapshot." + getUtcDateForComparison())
+        version.toString().contains('+testexample.')
```
Comment (OdysseusLives) [inline src/integTest/groovy/nebula/plugin/release/ReleasePluginIntegrationSpec.groovy]: Can you add a check that `+feature/testexample.` is not in the version.toString() to clarify this?

File: src/integTest/groovy/nebula/plugin/release/ReleasePluginIntegrationSpec.groovy [hunk 4]
```diff
@@ -776,6 +909,17 @@ class ReleasePluginIntegrationSpec extends GitVersioningIntegrationSpec {
...
 
+    def 'branches with dashes that do not match specified patterns do not fail builds - immutableSnapshot'() {
+        git.checkout(branch: 'dev-robtest', createBranch: true)
+
+        when:
+        def version = inferredVersionForTask('immutableSnapshot')
+
+        then:
+        version.toString().startsWith('0.1.0-snapshot.' + getUtcDateForComparison())
+        version.toString().contains('.robtest.')
```
Comment (OdysseusLives) [inline src/integTest/groovy/nebula/plugin/release/ReleasePluginIntegrationSpec.groovy]: This looks like it is different than what `devSnapshot` is doing in the previous test where `version == dev('0.1.0-dev.2+dev.robtest.')` Is that intentional?

File: README.md (hunk 1 above)
Comment (rpalcolea) [reply README.md]: Thanks for the suggestion, this is done now :)

File: README.md (hunk 2 above)
Comment (rpalcolea) [reply README.md]: I know! 👍

File: src/integTest/groovy/nebula/plugin/release/ReleasePluginIntegrationSpec.groovy (hunk 4 above)
Comment (rpalcolea) [reply src/integTest/groovy/nebula/plugin/release/ReleasePluginIntegrationSpec.groovy]: good catch, this is now fixed

File: src/integTest/groovy/nebula/plugin/release/ReleasePluginIntegrationSpec.groovy (hunk 3 above)
Comment (rpalcolea) [reply src/integTest/groovy/nebula/plugin/release/ReleasePluginIntegrationSpec.groovy]: good catch, this is now fixed

**Comment by rpalcolea** [comment]
Hi @OdysseusLives , let me know if  https://github.com/nebula-plugins/nebula-release-plugin/pull/153/commits/471dbd486c403f0beb5675a4e4e18cb88965341e makes sense to you

**Comment by rpalcolea** [comment]
Hey @OdysseusLives , based on our conversations and our findings in https://github.com/gradle/gradle/pull/9915

I believe this change is good to go, thoughts?

**Pull request merged**